### ISSUE TITLE
New UPE - Deferred payment intent integration

### ISSUE THREAD

**Issue by FangedParakeet** [Automattic/woocommerce-payments#5019]
### Description
Please refer to pc2DNy-2e2-p2 for more details.

> [Stripe has informed us that they have] added two important payment flow features with our desires in mind: the payment intent is no longer required prior to creating the payment method and the payment element can be used singularly to create the payment method, allowing us to create and confirm the payment intent ourselves on the Payments Server. This feature would solve our security concerns over exposed payment intent information on the client and would allow the UPE to be compatible with WooPay, since we can use our legacy payment intent confirmation flow with the UPE.
>
> Unfortunately, this feature is not yet ready. Stripe plans to open up a private beta for this feature by the end of October and plans to fully release the feature by the end of Q1 next year.

Please refer to [this file](https://wootransactp2.files.wordpress.com/2022/10/new-integration-path-for-payment-element.pdf.zip) for the current description of this new payment flow. 

If you scroll down to the section titled "Server-side confirmation", you will see that this almost exactly matches our current payment flow for our legacy card element, where the payment intent is created and confirmed simultaneously and subsequent to creating the payment method. This will allow us to resolve almost all of our current security and integration concerns with our existing UPE implementation. 

This issue exists to create the initial PoC implementation of this new payment flow, once the tool has been released to beta and we are given access to it by Stripe. This issue is simply to implement the basic shortcode checkout payment flow and there will be other issues created to support other checkout and payment flows. It's also assumed that this tool will be in a fluid state while in limited beta release, so this PoC will not need to be considered in a final state once completed. 

### Acceptance criteria

* Implements Enhanced UPE shortcode checkout flow.

### Additional context

[Stripe UPE – Options to solve WooPay compatibility and WC Pay loss of control over the payments UI/UX](pc2DNy-2e2-p2)
[New integration path for Payment Element](https://wootransactp2.files.wordpress.com/2022/10/new-integration-path-for-payment-element.pdf.zip)

**Comment by FangedParakeet** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Heisenberg-63586f2ec60359001310fa6a/issues/Automattic/woocommerce-payments/5019?planning-poker) with Zenhub @mdmoore @timur27

**Comment by timur27** [comment]
Do we want to use the new payment elements API to mount an element without creating payment intent in the legacy UPE or the split UPE? I thought the former, but I want to confirm this. Is the legacy UPE with the Stripe Payment Element and UPE card working together with WooPay enabled considered as a success of this issue?

Another question I want to consult with you all is if anyone thinks there is a need to update any code on the server side. [`create_intentions` method ](https://github.com/Automattic/woocommerce-payments-server/blob/trunk/server/wp-content/rest-api-plugins/endpoints/wcpay/class-intentions-controller.php#L281) looks promising, and there is a chance that no changes will be needed, but I might be wrong - could anyone agree or disagree? 

Although this is a POC, it still looks like a fairly large piece of work - can this POC be split into multiple pieces?

**Comment by FangedParakeet** [comment]
> Do we want to use the new payment elements API to mount an element without creating payment intent in the legacy UPE or the split UPE?

Yes! I'm not sure if there was a former or latter option posed here, but the goal is to use the new payment elements API to mount an element and we should not need to use any gateway to create a payment intent prior to mounting. 

> Is the legacy UPE with the Stripe Payment Element and UPE card working together with WooPay enabled considered as a success of this issue?

I think all split gateways should now be able to use the new payment elements API, so nothing should really need to work together per se. Obviously we don't want to break any other existing implementations, so I think this new UPE implementation will probably need to live behind yet another feature flag, so that all flavours of the UPE can coexist together, while not really being aware of each other's presence. 

> Another question I want to consult with you all is if anyone thinks there is a need to update any code on the server side. 

I am hopeful that this work can be completed without server-side updates, but I can't yet say this for sure. My dream is that the transaction flow will become similar to the legacy card flow, so much so that we can even reuse the `WC_Payment_Gateway_WCPay::process_payment` function, which eventually [creates and confirms a payment intent simultaneously](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payment-gateway-wcpay.php#L1094). 

This may also mean that most of our traditional UPE, AJAX checkout functions will become a thing of the past, for better or worse (probably better, though RIP). 

> Although this is a POC, it still looks like a fairly large piece of work - can this POC be split into multiple pieces?

I definitely agree here--both that this will be an unusually large effort and it would be nice to divide it. I can't think of obvious ways to divide the PoC, but I'm open to suggestions for any ways to collaborate more on our approach here. I do want to iterate as explicitly and loudly as possible that the goal of this PoC is *only* to **implement the simplest happy checkout flow for the card gateway**. 

Once this implementation is complete and the method to the madness decided, we can unblock all the other new/enhanced UPE issues referenced above and continue to complete the rest of this implementation together separately as a team.

**Comment by timur27** [comment]
> I'm not sure if there was a former or latter option posed here, but the goal is to use the new payment elements API to mount an element and we should not need to use any gateway to create a payment intent prior to mounting.

What I meant is that we currently are creating intents both in `upe-fields.js` and `upe-split-fields.js` prior to presenting the UI element component. At first glance, it looks like working with `upe-fields.js` only is the way to go, as split UPE will eventually be the first candidate to be thrown away, assuming the success of the enhanced UPE, and I see no big reason behind spending our efforts on `upe-split-fields.js` and everything related to it. I think that your statement `implement the simplest happy checkout flow for the card gateway.` confirms my train of thoughts, but please let me know in case I've missed anything crucial.


> I think this new UPE implementation will probably need to live behind yet another feature flag

This sounds reasonable! Do we want to introduce the flag as part of this issue? I haven't found anything related to the FF in other issues from the epic. 

> I am hopeful that this work can be completed without server-side updates, but I can't yet say this for sure. My dream is that the transaction flow will become similar to the legacy card flow

Got it! I'm with you here.

> I'm open to suggestions for any ways to collaborate more on our approach here

Given that `this PoC is only to implement the simplest happy checkout flow for the card gateway` as you mentioned, I see no real area for splitting this PoC, given the design of the code, which unfortunately won't allow us to work on the changes separately in parallel without creating conflicts that will slow us down.

I'll go and estimate now as I have everything needed and thanks for your patience while aligning on everything. Looking forward to hearing from you about the feature flag and the plans of introducing it (whether in this issue or not), but in the meantime, let me set the estimate already to not block us.

**Comment by FangedParakeet** [comment]
> What I meant is that we currently are creating intents both in upe-fields.js and upe-split-fields.js prior to presenting the UI element component. 

We will not be creating intents prior to presenting the payment element UI anymore.

> At first glance, it looks like working with upe-fields.js only is the way to go

This could be an option...though, I think this would be challenging, since I'm pretty sure [this entire section](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/blocks/upe-fields.js#L402-L422) will be redundant, since we will no longer require a payment intent at all on the client side. I sadly think that most of the front-end implementation--the scripts at least--will need to be recreated from scratch. 😞 

> I see no big reason behind spending our efforts on upe-split-fields.js and everything related to it

Indeed this has been a disappointingly temporary solution. 😞 x2

> Do we want to introduce the flag as part of this issue?

We could implement this first actually. I think you'd discussed creating some sort of feature flag factory or payment gateway/UPE factory before: I'm not sure if you got round to implementing that yet or not. I think we can definitely create a new issue for this--in fact, I just created #5682--and if you like, we can also create a new issue as well for any further refactors.

**Comment by timur27** [comment]
All the issues completed - closing the epic.
